File: WdtcCore/src/main/java/org/wdt/wdtc/core/download/fabric/FabricAPIDownloadTask.java
package org.wdt.wdtc.core.download.fabric;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import lombok.Getter;
import org.wdt.utils.gson.JsonArrayUtils;
import org.wdt.utils.gson.JsonObjectUtils;
import org.wdt.utils.io.FileUtils;
import org.wdt.wdtc.core.download.infterface.VersionJsonObjectInterface;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.utils.DownloadUtils;
import org.wdt.wdtc.core.utils.URLUtils;

import java.io.IOException;

public class FabricAPIDownloadTask {
  private static final String VersionListUrl = "https://api.modrinth.com/v2/project/P7dR8mSH/version";
  private final Launcher launcher;
  @Getter
  private final String FabricAPIVersionNumber;
  @Getter
  private VersionJsonObjectInterface versionJsonObjectInterface;

  public FabricAPIDownloadTask(Launcher launcher, String FabricAPIVersionNumber) {
    this.launcher = launcher;
    this.FabricAPIVersionNumber = FabricAPIVersionNumber;
  }

  public FabricAPIDownloadTask(Launcher launcher, VersionJsonObjectInterface versionJsonObjectInterface) {
    this(launcher, versionJsonObjectInterface.getVersionNumber());
    this.versionJsonObjectInterface = versionJsonObjectInterface;
  }

  public void startDownloadFabricAPI() throws IOException {
    FabricAPIVersionList.FabricAPIVersionJsonObjectImpl versionJsonObject = (FabricAPIVersionList.FabricAPIVersionJsonObjectImpl) versionJsonObjectInterface;
    if (versionJsonObject == null) {
      JsonArray VersionListArray = JsonArrayUtils.parseJsonArray(URLUtils.getURLToString(VersionListUrl));
      for (int i = 0; i < VersionListArray.size(); i++) {
        JsonObject VersionObject = VersionListArray.get(i).getAsJsonObject();
        FabricAPIVersionList.FabricAPIVersionJsonObjectImpl newVersionJsonObject =
            JsonObjectUtils.parseObject(VersionObject, FabricAPIVersionList.FabricAPIVersionJsonObjectImpl.class);
        if (newVersionJsonObject.getVersionNumber().equals(FabricAPIVersionNumber)) {
          downloadFabricAPITask(newVersionJsonObject.getFilesObjectList().get(0));
        }
      }
    } else {
      downloadFabricAPITask(versionJsonObject.getFilesObjectList().get(0));
    }
  }

  public void downloadFabricAPITask(FabricAPIVersionList.FabricAPIVersionJsonObjectImpl.FilesObject filesObject) {
    DownloadUtils.StartDownloadTask(filesObject.getJarDownloadURL(), FileUtils.toFile(launcher.getGameModsPath(), filesObject.getJarFileName()));
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/download/fabric/FabricAPIDownloadTask.kt
package org.wdt.wdtc.core.download.fabric

import org.wdt.utils.gson.parseJsonArray
import org.wdt.utils.gson.parseObject
import org.wdt.utils.io.toStrings
import org.wdt.wdtc.core.download.fabric.FabricAPIVersionList.FabricAPIVersionsJsonObjectImpl
import org.wdt.wdtc.core.download.fabric.FabricAPIVersionList.FabricAPIVersionsJsonObjectImpl.FilesObject
import org.wdt.wdtc.core.download.infterface.VersionsJsonObjectInterface
import org.wdt.wdtc.core.game.Version
import org.wdt.wdtc.core.utils.noNull
import org.wdt.wdtc.core.utils.startDownloadTask
import org.wdt.wdtc.core.utils.toURL
import java.io.File

class FabricAPIDownloadTask(private val version: Version, val fabricAPIVersionNumber: String?) {
	
	private var versionsJsonObjectInterface: VersionsJsonObjectInterface? = null
	private val versionListUrl = "https://api.modrinth.com/v2/project/P7dR8mSH/version".toURL()
	
	constructor(version: Version, versionsJsonObjectInterface: VersionsJsonObjectInterface) : this(
		version,
		versionsJsonObjectInterface.versionNumber
	) {
		this.versionsJsonObjectInterface = versionsJsonObjectInterface
	}
	
	suspend fun startDownloadFabricAPI() {
		val versionJsonObject = versionsJsonObjectInterface as FabricAPIVersionsJsonObjectImpl?
		if (versionJsonObject == null) {
			versionListUrl.toStrings().parseJsonArray().forEach {
				it.asJsonObject.parseObject<FabricAPIVersionsJsonObjectImpl>().also { api ->
					if (api.versionNumber == fabricAPIVersionNumber) {
						downloadFabricAPITask(api.filesObjectList.noNull().first())
					}
				}
			}
		} else {
			downloadFabricAPITask(versionJsonObject.filesObjectList.noNull().first())
		}
	}
	
	private suspend fun downloadFabricAPITask(filesObject: FilesObject) {
		startDownloadTask(filesObject.jarDownloadURL, File(version.gameModsPath, filesObject.jarFileName))
	}
	
	override fun toString(): String {
		return "FabricAPIDownloadTask(fabricAPIVersionNumber=$fabricAPIVersionNumber)"
	}
	
	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is FabricAPIDownloadTask) return false
		
		if (fabricAPIVersionNumber != other.fabricAPIVersionNumber) return false
		
		return true
	}
	
	override fun hashCode(): Int {
		return fabricAPIVersionNumber?.hashCode() ?: 0
	}
	
	
}
